File: KerTAD python/mask_1.py
import numpy as np
from triangle_method import *
#calculate first mask for an input matrix and return NxN binary matrix
def mask_1(input_map):
    #Calculate partial derivative using slicing (alternatively you can use scipy ndimage)
    Gy=np.zeros(input_map.shape)
    Gy[0:input_map.shape[1]-1,:]=input_map[1:input_map.shape[1],:] - input_map[0:-1,:]
    x=np.tile(np.ptp(Gy,1,keepdims=True), (1, Gy.shape[1]))
    x2=np.tile(np.sum(np.absolute(Gy),1,keepdims=True), (1, Gy.shape[1]))
    S=x*x.T* x2*x2.T
    #Go through diagonal of S and keep only local neighborhood maxima
    S2=np.zeros(S.shape)
    for i in range(1, S.shape[1]-1):
         nhood=S[i-1:i+2,i-1:i+2]
         if(S[i,i]==np.max(nhood)):
            S2[i,i]=1
         
    S2[0,:]=1
    S2[S2.shape[1]-1,:]=1

    binary_mask, threshold = triangle_method(S)
    S=S/np.max(S)
    M=((np.diag(S)>threshold)).reshape(-1, 1)*S2   
    M[0,0]=1;   
    M[-1,-1]=1;   
    d=np.expand_dims(np.diag(M),axis=1)
    L=np.tile(d, input_map.shape[0])
    M2=L+L.T
    empty_map=np.triu(M2>1)
    output_map=empty_map*((~np.eye(input_map.shape[0],dtype=bool)))
    first_mask=np.append(np.expand_dims(output_map[0,:],axis=0),np.zeros((1,output_map.shape[0])),axis=0)
    output_map=np.append(first_mask,output_map[1:-1,:],axis=0)
    return output_map

File: KerTAD/mask_1.m
%Calculates the first mask for a given input map. Returns binary NxN matrix
function [output_map]=mask_1(input_map)

%calculate partial derivative using imfilter
h=[-1 1];
Gy=imfilter(input_map,h','symmetric');
x=repmat(range(Gy,2), 1, size(Gy,2));
x2=repmat(sum(abs(Gy),2), 1, size(abs(Gy),2));
S=x.*x.'.* x2.*x2.';

%Filter local maxima
S2=zeros(size(S));
for i=2:size(S2,1)-1
         pt=[i i];
         idy = max(pt(1)-1,1):min(pt(1)+1,size(S,1));
         idx = max(pt(2)-1,1):min(pt(2)+1,size(S,2));
         nhood = S(idy,idx);
         if(S(i,i)==max(nhood,[],'all'))
            S2(i,i)=1;
         end

end
S2(1,:)=1; S2(size(S2,1),:)=1;
 

[~, threshold] = triangle_method (S);
S=S./max(S,[],'all');
M=double(imbinarize(diag(S./max(S(:))),threshold)).*S2;

M(1,1)=1; 
M(size(M,1),size(M,1))=1;
M2=repmat(diag(M),1, size(input_map,1)) + repmat(diag(M),1, size(input_map,1)).';

empty_map=triu(M2>1);
output_map=empty_map.*(double(~eye(size(input_map))));
first_mask=[output_map(1,:);zeros(1,size(output_map,1));output_map(2:end-1,:)];
output_map= first_mask; 

end